bigchaindb/util.py

```diff
@@ -175,52 +222,126 @@ def create_tx(current_owner, new_owner, tx_input, operation, payload=None):
+        if len(new_owners) > 1:
+            condition = cc.ThresholdSha256Fulfillment(threshold=len(new_owners))
+            for new_owner in new_owners:
+                condition.add_subfulfillment(cc.Ed25519Fulfillment(public_key=new_owner))
+        elif len(new_owners) == 1:
+            condition = cc.Ed25519Fulfillment(public_key=new_owners[0])
+        conditions.append({
+            'new_owners': new_owners,
+            'condition': {
+                'details': json.loads(condition.serialize_json()),
+                'uri': condition.condition.serialize_uri()
+            },
+            'cid': fulfillment['fid']
+        })
 
     tx = {
-        'current_owner': current_owner,
-        'new_owner': new_owner,
-        'input': tx_input,
+        'fulfillments': fulfillments,
+        'conditions': conditions,
         'operation': operation,
         'timestamp': timestamp(),
         'data': data
     }
 
     # serialize and convert to bytes
-    tx_serialized = serialize(tx)
-    tx_hash = crypto.hash_data(tx_serialized)
+    tx_hash = get_hash_data(tx)
 
     # create the transaction
     transaction = {
         'id': tx_hash,
+        'version': 1,
         'transaction': tx
     }
 
     return transaction
 
 
-def sign_tx(transaction, private_key):
+# TODO: Change sign_tx to populate the fulfillments
+def sign_tx(transaction, sks):
     """Sign a transaction
 
     A transaction signed with the `current_owner` corresponding private key.
 
     Args:
         transaction (dict): transaction to sign.
-        private_key (str): base58 encoded private key to create a signature of the transaction.
+        sk (base58 str): base58 encoded private key to create a signature of the transaction.
 
     Returns:
-        dict: transaction with the `signature` field included.
+        dict: transaction with the `fulfillment` fields populated.
 
     """
-    private_key = crypto.SigningKey(private_key)
-    signature = private_key.sign(serialize(transaction))
-    signed_transaction = transaction.copy()
-    signed_transaction.update({'signature': signature.decode()})
-    return signed_transaction
+    # validate sk
+    if not isinstance(sks, list):
+        sks = [sks]
+
+    # create a mapping between sk and vk so that we can match the private key to the current_owners
+    key_pairs = {}
+    for sk in sks:
+        signing_key = crypto.SigningKey(sk)
+        vk = signing_key.get_verifying_key().to_ascii().decode()
+        key_pairs[vk] = signing_key
+
+    tx = copy.deepcopy(transaction)
+
+    for fulfillment in tx['transaction']['fulfillments']:
+        fulfillment_message = get_fulfillment_message(transaction, fulfillment)
+        parsed_fulfillment = cc.Fulfillment.from_json(fulfillment_message['condition']['condition']['details'])
+        parsed_fulfillment_copy = copy.deepcopy(parsed_fulfillment)
+        # single current owner
+        if isinstance(parsed_fulfillment, cc.Ed25519Fulfillment):
```

Comment:
the branches in the `if` ... `else` can be put in smaller functions/methods to limit the complexity of the code and simplify tests.